File: douji-frontend/src/app/room/lib/SignalR/RoomStates/RoomState.ts
import { TimeProvider } from "@/app/lib/TimeProvider";

export const enum RoomStateEnum {
	Unstarted = 0,
	Playing = 1,
	Paused = 2,
	Waiting = 4,
	Ended = 5,
}

export class RoomState {
	public constructor(
		public readonly state: RoomStateEnum,
		public readonly videoTime: number | null,
		public readonly updatedAt: Date
	) {}

	public getCurrentExpectedTime(): number | null {
		if (this.videoTime == null)
			return null;

		return ((TimeProvider.now() - this.updatedAt.getTime()) / 1000) + this.videoTime;
	}
}

export class RoomStateDTO {
	public constructor(
		public readonly state: RoomStateEnum,
		public readonly videoTime: number | null,
		public readonly updatedAt: string
	) {}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	public static fromObject(obj: any): RoomStateDTO {
		return new RoomStateDTO(obj["state"], obj["videoTime"], obj["updatedAt"]);
	}

	public toRoomState(): RoomState {
		return new RoomState(this.state, this.videoTime, new Date(this.updatedAt));
	}
}

File: Douji Backend/Model/RoomStates/RoomState.cs
using Douji.Backend.Model.ClientStates;

namespace Douji.Backend.Model.RoomStates;

public enum RoomStateEnum
{
	Unstarted = 0,
	Playing = 1,
	Paused = 2,
	Waiting = 4,
	Ended = 5,
}

public abstract class RoomState(RoomStateEnum state, double? videoTime, DateTime updatedAt)
{
	public RoomStateEnum State { get; } = state;
	public double? VideoTime { get; } = videoTime;
	public DateTime UpdatedAt { get; } = updatedAt;

	public abstract RoomState? AcceptClientStateEvent(ClientState clientState, User user, Room room);
}

public abstract class RoomStateWithTime(RoomStateEnum state, double videoTime, DateTime updatedAt)
	: RoomState(state, videoTime, updatedAt)
{
	public double VideoTimeGuaranteed { get; } = videoTime;

	public abstract double GetCurrentExpectedTime();
}
